### ISSUE TITLE
Review blockstack transaction generation API and plans for integration

### ISSUE THREAD

**Issue by kantai** [stx-labs/stacks.js#348]
As implemented, the method signatures for transaction generation in `blockstack.js` look like:

```javascript
function makePreorder(fullyQualifiedName: string,
                      destinationAddress: string,
                      paymentKey: bitcoinjs.ECPair,
                      network: BlockstackNetwork): Promise<bitcoinjs.Transaction>
```
```javascript
function makeRegister(fullyQualifiedName: string,
                             registerAddress: string,
                             paymentKey: bitcoinjs.ECPair,
                             zonefile: string = null,
                             network: BlockstackNetwork): Promise<bitcoinjs.Transaction>
```

There's a couple immediate issues here --

1. Parameters are tied to implementation detail (bitcoinjs objects) -- they should accept either WIFs or private-key hexes (I prefer WIFs, as they are somewhat standardized in the ecosystem, while private-key hex strings _aren't_)
2. All this does is generate a transaction -- browser currently just performs an API call which generates the transactions and performs the broadcast. This means that the integration would have to change somewhat significantly.
3. blockstack-core will coerce given addresses (and responses) to mainnet addresses / testnet addresses, depending on which network it is running on. Should `blockstack.js` do the same?
4. the network option comes from the classes defined here: https://github.com/blockstack/blockstack.js/blob/feature/blockstack-operations/src/operations/network.js and there's some supplied constants for default configuration. Is this the best way to configure and pass the network configuration?

Based on the answers to the above, we can discuss how this should be integrated into blockstack-browser, which will have to (1) broadcast transactions (2) maintain the network configuration settings and (3) potentially use a transaction broadcast _service_ to broadcast the REGISTER operation.

**Comment by kantai** [comment]
Tagging @jcnelson @larrysalibra @yknl for input on the API and integration plans.

**Comment by jcnelson** [comment]
The way blockstack-core (and all of its libraries) figures out which network to use is by checking an environment variable.  Would this be a reasonable thing to do in the browser?

One issue to consider is safety checks.  For example, when registering a name, the transaction should not be sent if the name is already registered by someone else.  Do we want blockstack.js to do these checks?  If we don't, then an alternative is to have transaction broadcast API endpoint check transactions (at the caller's option).  It could simply recycle the safety-check code we have in Python.

> (3) potentially use a transaction broadcast service to broadcast the REGISTER operation.

By this, do you mean a stateful network service that relays your REGISTER transaction at the right time, once your PREORDER is accepted?

**Comment by kantai** [comment]
> The way blockstack-core (and all of its libraries) figures out which network to use is by checking an environment variable. Would this be a reasonable thing to do in the browser?

Probably not -- the browser runs as a basically serverless webapp -- which means it wouldn't necessarily have access to environment variables. If we were pairing it with a native service, then we'd be able to go that route.

> One issue to consider is safety checks. For example, when registering a name, the transaction should not be sent if the name is already registered by someone else. Do we want blockstack.js to do these checks? If we don't, then an alternative is to have transaction broadcast API endpoint check transactions (at the caller's option). It could simply recycle the safety-check code we have in Python.

Yeah -- I think these safety checks should be in `blockstack.js`

> By this, do you mean a stateful network service that relays your REGISTER transaction at the right time, once your PREORDER is accepted?

Yep -- the user would be trusting that service not to front-run their registrations. We could present the user with two options (1) use a trusted stateful relay or (2) display the transaction as a QR code or hex string, and say "wait an hour and then broadcast this on your favorite bitcoin transaction broadcaster".

**Comment by kantai** [comment]
From today's engineering meeting:

1. Let's continue to use private-key hex-strings -- this is a standard from OpenSSL. WIFs are a little too bitcoin-centric
2. We should _stop_ coercing addresses -- this was a stop-gap measure in blockstack-browser because it doesn't know how to deal with testnet addresses, however, we should generally use testnet addresses when on testnet.
3. These configuration variables should actually be in the top-level blockstack.js object

**Comment by kantai** [comment]
Per item three in last week's engineering meeting -- we have some amount of configuration data for `blockstack.js`, and that's currently stored in localStorage. This is bad -- it forces applications to store their configuration data in a specific way, and it breaks webworkers. The proposed alternative was to store the configuration settings as quasi-globals in the `blockstack` module.

Now, towards that end -- in the `feature/operations` branch, where I've been working on implementing transaction generation in the client library -- I added a `config` object. This starts with the default network options, but these can be changed:

```javascript
    let bsk = require('blockstack')
    // default local regtest configuration
    bsk.config.network = bsk.network.defaults.LOCAL_REGTEST
    // default mainnet (which is also the default on first import)
    bsk.config.network = bsk.network.defaults.MAINNET_DEFAULT
    // specific utxoProvider and blockstackAPI endpoint
    let utxoProvider = 'https://blockchain.info/unspent?format=json&active='
    let blockstackAPI = 'https://core.blockstack.org'
    bsk.config.network = new bsk.network.BlockstackNetwork(blockstackAPI, utxoProvider)
```

I would appreciate feedback on the above -- one note: the reason I've defined network objects instead of just using URLs as configuration variables is that different network will require different communication behaviors -- local regtest broadcasts transactions and queries UTXOs differently from deployed APIs.

This is also related to issues #235 and #224
